ext/ddtrace_profiling_native_extension/collectors_discrete_dynamic_sampler.c

```diff
@@ -0,0 +1,323 @@
+  //       changes (e.g. probabilities in the range of ]50%, 100%[ which map to intervals in the range of ]1, 2[). Our approach
+  //       when the sampling interval is a non-integer is to ceil it (i.e. we'll always choose to sample less often).
+  sampler->sampling_interval = ceil(1.0 / sampler->sampling_probability);
+
+  #ifdef DD_DEBUG
+    double allocs_in_60s = sampler->events_per_ns * 1e9 * 60;
+    double samples_in_60s = allocs_in_60s * sampler->sampling_probability;
+    double expected_total_sampling_time_in_60s =
+      samples_in_60s * sampler->sampling_time_ns / 1e9;
+    double real_total_sampling_time_in_60s = sampling_window_time_ns / 1e9 * 60 / (window_time_ns / 1e9);
+
+    fprintf(stderr, "[dds.%s] readjusting...\n", sampler->id);
+    fprintf(stderr, "samples_since_last_readjustment=%ld\n", sampler->samples_since_last_readjustment);
+    fprintf(stderr, "window_time=%ld\n", window_time_ns);
+    fprintf(stderr, "events_per_sec=%f\n", sampler->events_per_ns * 1e9);
+    fprintf(stderr, "sampling_time=%ld\n", sampler->sampling_time_ns);
+    fprintf(stderr, "sampling_window_time=%ld\n", sampling_window_time_ns);
+    fprintf(stderr, "sampling_overshoot_time=%ld\n", sampling_overshoot_time_ns);
+    fprintf(stderr, "working_window_time=%ld\n", working_window_time_ns);
+    fprintf(stderr, "sampling_interval=%zu\n", sampler->sampling_interval);
+    fprintf(stderr, "sampling_probability=%f\n", sampler->sampling_probability);
+    fprintf(stderr, "expected allocs in 60s=%f\n", allocs_in_60s);
+    fprintf(stderr, "expected samples in 60s=%f\n", samples_in_60s);
+    fprintf(stderr, "expected sampling time in 60s=%f (previous real=%f)\n", expected_total_sampling_time_in_60s, real_total_sampling_time_in_60s);
+    fprintf(stderr, "target_overhead_adjustment=%f\n", sampler->target_overhead_adjustment);
+    fprintf(stderr, "expected max overhead in 60s=%f\n", target_overhead / 100.0 * 60);
+    fprintf(stderr, "-------\n");
+  #endif
+
+  sampler->events_since_last_readjustment = 0;
+  sampler->samples_since_last_readjustment = 0;
+  sampler->sampling_time_since_last_readjustment_ns = 0;
+  sampler->last_readjust_time_ns = now;
+}
+
+// ---
+// Below here is boilerplate to expose the above code to Ruby so that we can test it with RSpec as usual.
+
+static VALUE _native_new(VALUE klass);
+static VALUE _native_reset(VALUE self, VALUE target_overhead);
+static VALUE _native_should_sample(VALUE self, VALUE now);
+static VALUE _native_after_sample(VALUE self, VALUE now);
+static VALUE _native_probability(VALUE self);
+
+void collectors_discrete_dynamic_sampler_init(VALUE profiling_module) {
+  VALUE collectors_module = rb_define_module_under(profiling_module, "Collectors");
+  VALUE testing_module = rb_define_module_under(collectors_module, "Testing");
+  VALUE sampler_class = rb_define_class_under(testing_module, "DiscreteDynamicSampler", rb_cObject);
```

Comment:
Hmm yeah I see what you mean. My initial intention was to restrict the exposure of `DiscreteDynamicSampler` to Ruby for specs to prevent people wanting to re-use it in Ruby side and introducing Ruby<>Native hand-offs for every event they wanted to sample.

Also since a lot of the specs on it are very "continuous" in nature, I didn't feel the approach taken by the `DynamicSamplingRate` with its "pure" stateless native functions was correct either.

I guess I could also go with `Profiling::Collectors::DiscreteDynamicSampler::Testing::Sampler` or some variation on that to prevent ending up with weird `DiscreteDynamicSampler::Testing::DiscreteDynamicSampler`?

Reply:
Ah right, because these are instance methods and not static methods. I kinda saw it, but didn't make the connection...

I think `Testing::Sampler` would be fine? Or we could make `DiscreteDynamicSampler::Testing` itself a class -- like other Ruby classes, you can have static methods on it, or just create new instances.